import pandas as pd
sample_df = pd.read_table("PutahCreek_samples.txt", header = None, names = ['ID']) # Relevant data provided in Table S2 of associated manuscript
SAMPLES = sample_df['ID'].tolist()


# Local paths to genome assemblies. Genome assemblies have been made public on NCBI, accession numbers provided in the associated manuscript. 
RAWLINS_HAP1="/group/gmcoopgrp2/jgroh/heterodichogamy2/HiFi_assemblies/Rawlins/bcov2/Rawlins_hap1.fasta"
RAWLINS_HAP1_FAI="/group/gmcoopgrp2/jgroh/heterodichogamy2/HiFi_assemblies/Rawlins/bcov2/Rawlins_hap1.fasta.fai"

RAWLINS_HAP2="/group/gmcoopgrp2/jgroh/heterodichogamy2/HiFi_assemblies/Rawlins/bcov2/Rawlins_hap2.fasta"
RAWLINS_HAP2_FAI="/group/gmcoopgrp2/jgroh/heterodichogamy2/HiFi_assemblies/Rawlins/bcov2/Rawlins_hap2.fasta.fai"

# Earlier version of the manuscript ran all analyses aligning to J. californica genomes
Jcali_primary_fai = pd.read_table("/home/jgroh/heterodichogamy/Juglans_genome_assemblies/Jcalifornica/Jcali_primary.fna.fai", header = None, names = ['chr', 'len', 'V1', 'V2', 'V3'])
JCALI_P_CHRS = Jcali_primary_fai['chr'].tolist()

Rawlins_fai = pd.read_table("/group/gmcoopgrp2/jgroh/heterodichogamy2/HiFi_assemblies/Rawlins/bcov2/Rawlins_hap2.fasta.fai", header = None, names = ['chr', 'len', 'V1', 'V2', 'V3'])
RAWLINS_CHRS = Rawlins_fai['chr'].tolist()

Jcali_alt_fai = pd.read_table("/home/jgroh/heterodichogamy/Juglans_genome_assemblies/Jcalifornica/Jcali_alt.fna.fai", header = None, names = ['chr', 'len', 'V1', 'V2', 'V3'])
JCALI_A_CHRS = Jcali_alt_fai['chr'].tolist()


# These are samples from Stevens et al. 2018 (full citation in associated manuscript)
OLDSAMPLES = ['JHIN01', 'JHIN02', 'JHIN03', 'JHIN04', 'JHIN05', 'JHIN06', 'JHIN08', 'JHIN09', 'JHIN10', 'JHIN11']
# JHIN07 appears to be the same sequence as JHIN01, exclude

CHROMS = ['01','02','03','04','05','06','07','08','09','10','11','12','13','14','15','16']

JMIC_GENOME = '/home/jgroh/heterodichogamy/Jmicrocarpa_31.01_pseudohaploid_assembly/GCF_004785595.1_Jm3101_v1.0_genomic.fna'
JMIC_BWA_INDEX = '/home/jgroh/heterodichogamy/Jmicrocarpa_31.01_pseudohaploid_assembly/GCF_004785595.1_Jm3101_v1.0_genomic.fna.bwt'
JMIC_SAMTOOLS_INDEX = '/home/jgroh/heterodichogamy/Jmicrocarpa_31.01_pseudohaploid_assembly/GCF_004785595.1_Jm3101_v1.0_genomic.fna.fai'

GENOMES = ['Jcali_primary', 'Jcali_alt']

SAMTOOLS="/home/jgroh/heterodichogamy/conda_envs/samtools_1.17.yaml"

HIFI_DIR="/group/gmcoopgrp2/jgroh/heterodichogamy2/HiFi_assemblies/Rawlins/bcov2/"

HINDSII_GENOMES = {
    'Rawlins_hap1': HIFI_DIR+'Rawlins_hap1.fasta',
    'Rawlins_hap2': HIFI_DIR+'Rawlins_hap2.fasta'
    }


# local path to resequencing files.
DATA_DIR="/group/gmcoopgrp2/jgroh/heterodichogamy2/hindsii/"


# These define the output files. Files can be generated by uncommenting lines below. Note it is not recommended to run the entire pipeline at once.
rule all:
  input:
#    "output/Putah_Rawlins_hap2.assoc.txt",
#    "output/Putah_Rawlins_hap2_lmm1.assoc.txt",
#    "calls/Rawlins_hap2/Putah_44inds.fam",
#    expand("results/coverage_Rawlins/{sample}_Rawlins_hap1_Gloc.txt.gz", sample = SAMPLES),
#    expand(DATA_DIR+"alignment_files/Rawlins_hap{N}/{sample}.cram.crai", N = ['1', '2'], sample = SAMPLES),
#    expand("Putah_Gloc_{gt}_2Jcali_{assembly}.geno.ld", gt = ['all','protandrous'], assembly = ['primary', 'alt']),
#    "output/Putah2JcaliP.assoc.txt",
#    "output/Putah2JcaliAlt.assoc.txt",
#    expand("calls/{genome}/Putah_allsites_filtered_{genome}.vcf.gz", genome = ['Rawlins_hap2']),
#    expand('results/coverage_{genome}/{sample}.txt.gz', genome=GENOMES, sample = SAMPLES),
#    expand('results/coverage_{genome}/{sample}_norm.txt', genome=GENOMES, sample = SAMPLES),
    

rule index_Rawlins:
  input:
    '/group/gmcoopgrp2/jgroh/heterodichogamy2/HiFi_assemblies/Rawlins/bcov2/{gnom}.fasta'
  output:
    '/group/gmcoopgrp2/jgroh/heterodichogamy2/HiFi_assemblies/Rawlins/bcov2/{gnom}.fasta.bwt'
  conda:
    SAMTOOLS
  shell:
    "bwa index {input}"
    
rule bwa_map_to_Rawlins:
  input:
    fa=lambda wildcards: HINDSII_GENOMES[wildcards.genome],
    bwt=lambda wildcards: HINDSII_GENOMES[wildcards.genome]+'.bwt',
    r1="fastq/{sample}_R1.fq.gz",
    r2="fastq/{sample}_R2.fq.gz",
  output:
    DATA_DIR+"alignment_files/{genome}/{sample}.cram"
  conda:
    SAMTOOLS
  resources:
    runtime = 24*60,
    mem_mb=30000
  threads:
    24
  shell:
    """
    mkdir -p /group/gmcoopgrp2/jgroh/heterodichogamy2/hindsii/tmp_sorted_reads/{wildcards.genome}/{wildcards.sample} &&
    bwa mem -t {threads} -M \
        -R '@RG\\tID:{wildcards.sample}\\tLB:{wildcards.sample}\\tSM:{wildcards.sample}\\tPL:illumina' \
        {input.fa} {input.r1} {input.r2} | \
        samtools sort -O bam -l 0 -T /group/gmcoopgrp2/jgroh/heterodichogamy2/hindsii/tmp_sorted_reads/{wildcards.genome}/{wildcards.sample} - | \
        samtools view -T {input.fa} -C -o {output} -
    rm -r /group/gmcoopgrp2/jgroh/heterodichogamy2/hindsii/tmp_sorted_reads/{wildcards.genome}/{wildcards.sample}
    """

rule bwa_map_to_californica:
  input:
    fa="/home/jgroh/heterodichogamy/Juglans_genome_assemblies/Jcalifornica/{genome}.fna",
    index="/home/jgroh/heterodichogamy/Juglans_genome_assemblies/Jcalifornica/{genome}.fna.bwt",
    r1="fastq/{sample}_R1.fq.gz",
    r2="fastq/{sample}_R2.fq.gz",
  output:
    "alignment_files/{genome}/{sample}.cram"
  conda:
    "/home/jgroh/heterodichogamy/conda_envs/samtools_1.17.yaml"
  resources:
    runtime = 24*60,
    mem_mb=30000
  threads:
    24
  shell:
    """
    mkdir -p /group/gmcoopgrp2/jgroh/heterodichogamy2/hindsii/tmp_sorted_reads/{wildcards.genome}/{wildcards.sample} &&
    bwa mem -t {threads} -M \
        -R '@RG\\tID:{wildcards.sample}\\tLB:{wildcards.sample}\\tSM:{wildcards.sample}\\tPL:illumina' \
        {input.fa} {input.r1} {input.r2} | \
        samtools sort -O bam -l 0 -T /group/gmcoopgrp2/jgroh/heterodichogamy2/hindsii/tmp_sorted_reads/{wildcards.genome}/{wildcards.sample} - | \
        samtools view -T {input.fa} -C -o {output} -
    rm -r /group/gmcoopgrp2/jgroh/heterodichogamy2/hindsii/tmp_sorted_reads/{wildcards.genome}/{wildcards.sample}
    """

rule index_cram_Rawlins:
  input:
    DATA_DIR+"alignment_files/{genome}/{sample}.cram"
  output:
    DATA_DIR+"alignment_files/{genome}/{sample}.cram.crai"
  conda:
    SAMTOOLS
  shell:
    "samtools index {input}"

rule index_cram_californica:
  input:
    "alignment_files/{genome}/{sample}.cram"
  output:
    "alignment_files/{genome}/{sample}.cram.crai"
  conda:
    "/home/jgroh/heterodichogamy/conda_envs/samtools_1.17.yaml"
  shell:
    "samtools index {input}"

rule samtools_coverage_Rawlins:
  input:
    cram_hap1="/group/gmcoopgrp2/jgroh/heterodichogamy2/hindsii/alignment_files/Rawlins_hap1/{sample}.cram",
    crai_hap1="/group/gmcoopgrp2/jgroh/heterodichogamy2/hindsii/alignment_files/Rawlins_hap1/{sample}.cram.crai",
    cram_hap2="/group/gmcoopgrp2/jgroh/heterodichogamy2/hindsii/alignment_files/Rawlins_hap2/{sample}.cram",
    crai_hap2="/group/gmcoopgrp2/jgroh/heterodichogamy2/hindsii/alignment_files/Rawlins_hap2/{sample}.cram.crai",
  output:
    hap1_focal="results/coverage_Rawlins/{sample}_Rawlins_hap1_Gloc.txt.gz",
    hap2_focal="results/coverage_Rawlins/{sample}_Rawlins_hap2_Gloc.txt.gz",
  conda:
    "/home/jgroh/heterodichogamy/conda_envs/samtools_1.17.yaml"
  shell:
    """
    samtools depth -a {input.cram_hap1} -r JAKSXK010000007.1_RagTag:30660000-30740000 | gzip > {output.hap1_focal}
    samtools depth -a {input.cram_hap2} -r JAKSXK010000007.1_RagTag:32295000-32375000 | gzip > {output.hap2_focal}
    """

#avg="results/coverage_Rawlins/{sample}_genome_avg.txt"
#samtools depth -a {input.cram_hap1} | awk '{{ sum += $3 }} END {{ mean = sum / NR; print mean }}' > {output.avg}

rule samtools_coverage_Jcali_primary:
  input:
    cram="alignment_files/Jcali_primary/{sample}.cram",
    crai="alignment_files/Jcali_primary/{sample}.cram.crai"
  output:
    focal="results/coverage_Jcali_primary/{sample}.txt.gz",
    norm="results/coverage_Jcali_primary/{sample}_norm.txt"
  conda:
    "/home/jgroh/heterodichogamy/conda_envs/samtools_1.17.yaml"
  shell:
    """
    samtools depth -a {input.cram} -r JAKSXK010000007.1:31335000-31450000 | gzip > {output.focal}
    samtools depth -a {input.cram} -r JAKSXK010000001.1 | awk '{{ sum += $3 }} END {{ mean = sum / NR; print mean }}' > {output.norm}
    """


rule samtools_coverage_Jcali_alt:
  input:
    cram="alignment_files/Jcali_alt/{sample}.cram",
    crai="alignment_files/Jcali_alt/{sample}.cram.crai"
  output:
    focal="results/coverage_Jcali_alt/{sample}.txt.gz",
    norm="results/coverage_Jcali_alt/{sample}_norm.txt"
  conda:
    "/home/jgroh/heterodichogamy/conda_envs/samtools_1.17.yaml"
  shell:
    """
    samtools depth -a {input.cram} -r JAKSXL010000006.1:30490000-30536000 | gzip > {output.focal}
    samtools depth -a {input.cram} -r JAKSXL010000001.1 | awk '{{ sum += $3 }} END {{ mean = sum / NR; print mean }}' > {output.norm}
    """

rule bcftools_call_by_chrom: 
  input:
    fa="/group/gmcoopgrp2/jgroh/heterodichogamy2/HiFi_assemblies/Rawlins/bcov2/Rawlins_hap2.fasta",
    cram=expand("/group/gmcoopgrp2/jgroh/heterodichogamy2/hindsii/alignment_files/Rawlins_hap2/{sample}.cram", sample = SAMPLES),
    crai=expand("/group/gmcoopgrp2/jgroh/heterodichogamy2/hindsii/alignment_files/Rawlins_hap2/{sample}.cram.crai", sample = SAMPLES)
  output:
    "calls/Rawlins_hap2/chrs/{chr}.vcf.gz"
  resources:
    runtime= 72*60,
    mem_mb=60000
  conda:
    "/home/jgroh/heterodichogamy/conda_envs/samtools_1.17.yaml"
  shell:
    """
    bcftools mpileup -Ou -f {input.fa} -r {wildcards.chr} \
        --annotate "AD,DP,INFO/AD" --max-depth 200 {input.cram} | bcftools call -m -Oz -f GQ,GP -o {output}
    """

rule bcftools_concat_chr_vcfs_Rawlins_hap2:
  input:
    expand("calls/Rawlins_hap2/chrs/{chr}.vcf.gz", chr = RAWLINS_CHRS)
  output:
    "calls/Rawlins_hap2/Putah_allsites_unfiltered.vcf.gz"
  conda:
    "/home/jgroh/heterodichogamy/conda_envs/samtools_1.17.yaml"
  shell:
    "bcftools concat {input} -o {output}"


rule bcftools_concat_chr_vcfs_Jcali_primary:
  input:
    expand("calls/Jcali_primary/chrs/{chr}.vcf.gz", chr = JCALI_P_CHRS)
  output:
    "calls/Jcali_primary/Putah_allsites_unfiltered.vcf.gz"
  conda:
    "/home/jgroh/heterodichogamy/conda_envs/samtools_1.17.yaml"
  shell:
    "bcftools concat {input} -o {output}"

rule bcftools_concat_chr_vcfs_Jcali_alt:
  input:
    expand("calls/Jcali_alt/chrs/{chr}.vcf.gz", chr = JCALI_A_CHRS)
  output:
    "calls/Jcali_alt/Putah_allsites_unfiltered.vcf.gz"
  conda:
    "/home/jgroh/heterodichogamy/conda_envs/samtools_1.17.yaml"
  shell:
    "bcftools concat {input} -o {output}"



rule vcftools_filter_variant_sites:
  input:
    "calls/{genome}/Putah_allsites_unfiltered.vcf.gz"
  output:
    "calls/{genome}/Putah_variant_filtered.vcf.gz"
  conda:
    "/home/jgroh/heterodichogamy/conda_envs/vcftools.yaml"
  shell:
    """
    vcftools --gzvcf {input} \
        --remove-indels \
        --mac 2 \
        --max-missing 0.2 \
        --minGQ 30 \
        --minDP 10 \
        --maxDP 200 \
        --recode --stdout | bgzip -c > {output}
    """

rule vcftools_filter_invariant_sites:
  input:
    "calls/{genome}/Putah_allsites_unfiltered.vcf.gz"
  output:
    "calls/{genome}/Putah_invariant_filtered.vcf.gz"
  conda:
    "/home/jgroh/heterodichogamy/conda_envs/vcftools.yaml"
  resources:
    runtime=24*60,
    mem_mb=10000
  shell:
    """
    vcftools --gzvcf {input} \
        --remove-indels \
        --max-maf 0 \
        --max-missing 0.2 \
        --minGQ 30 \
        --minDP 10 \
        --maxDP 200 \
        --recode --stdout | bgzip -c > {output}
    """

rule index_filtered_vcfs:
  input:
    "calls/{genome}/Putah_{sites}_filtered.vcf.gz"
  output:
    "calls/{genome}/Putah_{sites}_filtered.vcf.gz.csi"
  conda:
    "/home/jgroh/heterodichogamy/conda_envs/samtools_1.17.yaml"
  shell:
    "bcftools index {input}"


rule concat_vcfs:
  input:
    variant="calls/{genome}/Putah_variant_filtered.vcf.gz",
    variant_csi="calls/{genome}/Putah_variant_filtered.vcf.gz.csi",
    invariant="calls/{genome}/Putah_invariant_filtered.vcf.gz",
    invariant_csi="calls/{genome}/Putah_invariant_filtered.vcf.gz.csi",
  output:
    "calls/{genome}/Putah_allsites_filtered_{genome}.vcf.gz"
  conda:
    "/home/jgroh/heterodichogamy/conda_envs/samtools_1.17.yaml"
  shell:
    "bcftools concat --allow-overlaps {input.variant} {input.invariant} -O z -o {output}"


rule keep_ids:
  input:
    variant="calls/{genome}/Putah_variant_filtered.vcf.gz",
  output:
    temp("calls/{genome}/Putah_variant_filtered_{genome}_44ids.bcf")
  conda:
    "/home/jgroh/heterodichogamy/conda_envs/samtools_1.17.yaml"
  shell:
    """
    bcftools view --samples ^JHIN_PC_040,JHIN_PC_060,JHIN_PC_052,JHIN_PC_102 -Ou -o {output} {input}
    """

##### RUN GWAS #####
rule vcf_to_plink_Rawlins_hap2:
  input:
    "calls/Rawlins_hap2/Putah_variant_filtered_Rawlins_hap2_44ids.bcf"
  output:
    bed="calls/Rawlins_hap2/Putah_44inds.bed",
    bim="calls/Rawlins_hap2/Putah_44inds.bim",
    fam="calls/Rawlins_hap2/Putah_44inds.fam",
  conda:
    "/home/jgroh/heterodichogamy/conda_envs/plink_1.90.yaml"
  shell:
    "plink --bcf {input} --const-fid 0 --allow-extra-chr --out calls/Rawlins_hap2/Putah_44inds"

rule vcf_to_plink_JcaliAlt:
  input:
    "calls/Jcali_alt/Putah_variant_filtered_Jcali_alt_44ids.bcf"
  output:
    bed="calls/Jcali_alt/Putah2JcaliAlt.bed",
    bim="calls/Jcali_alt/Putah2JcaliAlt.bim",
    fam="calls/Jcali_alt/Putah2JcaliAlt.fam",
  conda:
    "/home/jgroh/heterodichogamy/conda_envs/plink_1.90.yaml"
  shell:
    "plink --bcf {input} --const-fid 0 --allow-extra-chr --out calls/Jcali_alt/Putah2JcaliAlt"

rule vcf_to_plink_JcaliPrimary:
  input:
    "calls/Jcali_primary/Putah_variant_filtered_Jcali_primary_44ids.bcf"
  output:
    bed="calls/Jcali_primary/Putah2JcaliP.bed",
    bim="calls/Jcali_primary/Putah2JcaliP.bim",
    fam="calls/Jcali_primary/Putah2JcaliP.fam",
  conda:
    "/home/jgroh/heterodichogamy/conda_envs/plink_1.90.yaml"
  shell:
    "plink --bcf {input} --const-fid 0 --allow-extra-chr --out calls/Jcali_primary/Putah2JcaliP"

# * manual edit of fam with correct phenotype information. Phenotypes provided in Table S2 of associated manuscript. *

rule gemma_relatedness_matrix_Rawlins_hap2:
  input:
    bed="calls/Rawlins_hap2/Putah_44inds.bed",
    bim="calls/Rawlins_hap2/Putah_44inds.bim",
    fam="calls/Rawlins_hap2/Putah_44inds.fam", #AFTER EDIT
  output:
    "output/Putah_Rawlins_hap2.cXX.txt"
  conda:
    "/home/jgroh/heterodichogamy/conda_envs/gemma.yaml"
  shell:
    "gemma -bfile calls/Rawlins_hap2/Putah_44inds -gk 1 -o Putah_Rawlins_hap2"

rule gemma_relatedness_matrix_primary:
  input:
    bed="calls/Jcali_primary/Putah2JcaliP.bed",
    bim="calls/Jcali_primary/Putah2JcaliP.bim",
    fam="calls/Jcali_primary/Putah2JcaliP.fam", #AFTER EDIT
  output:
    "output/Putah2JcaliP.cXX.txt"
  conda:
    "/home/jgroh/heterodichogamy/conda_envs/gemma.yaml"
  shell:
    "gemma -bfile calls/Jcali_primary/Putah2JcaliP -gk 1 -o Putah2JcaliP"


rule gemma_relatedness_matrix_alt:
  input:
    bed="calls/Jcali_alt/Putah2JcaliAlt.bed",
    bim="calls/Jcali_alt/Putah2JcaliAlt.bim",
    fam="calls/Jcali_alt/Putah2JcaliAlt.fam", #AFTER EDIT
  output:
    "output/Putah2JcaliAlt.cXX.txt"
  conda:
    "/home/jgroh/heterodichogamy/conda_envs/gemma.yaml"
  shell:
    "gemma -bfile calls/Jcali_alt/Putah2JcaliAlt -gk 1 -o Putah2JcaliAlt"


rule gemma_lmm_Rawlins_hap2:
  input:
    bed="calls/Rawlins_hap2/Putah_44inds.bed",
    bim="calls/Rawlins_hap2/Putah_44inds.bim",
    fam="calls/Rawlins_hap2/Putah_44inds.fam", #AFTER EDIT
    mat="output/Putah_Rawlins_hap2.cXX.txt"
  output:
    "output/Putah_Rawlins_hap2.assoc.txt"
  conda:
    "/home/jgroh/heterodichogamy/conda_envs/gemma.yaml"
  resources:
    runtime=60,
    mem_mb=50000
  shell:
    "gemma -bfile calls/Rawlins_hap2/Putah_44inds -k {input.mat} -lmm 2 -miss 0.2 -o Putah_Rawlins_hap2"

rule gemma_lmm_Rawlins_hap2_lmm1:
  input:
    bed="calls/Rawlins_hap2/Putah_44inds.bed",
    bim="calls/Rawlins_hap2/Putah_44inds.bim",
    fam="calls/Rawlins_hap2/Putah_44inds.fam", #AFTER EDIT
    mat="output/Putah_Rawlins_hap2.cXX.txt"
  output:
    "output/Putah_Rawlins_hap2_lmm1.assoc.txt"
  conda:
    "/home/jgroh/heterodichogamy/conda_envs/gemma.yaml"
  resources:
    runtime=60,
    mem_mb=50000
  shell:
    "gemma -bfile calls/Rawlins_hap2/Putah_44inds -k {input.mat} -lmm 1 -miss 0.2 -o Putah_Rawlins_hap2_lmm1"

rule gemma_lmm_primary:
  input:
    bed="calls/Jcali_primary/Putah2JcaliP.bed",
    bim="calls/Jcali_primary/Putah2JcaliP.bim",
    fam="calls/Jcali_primary/Putah2JcaliP.fam", #AFTER EDIT
    mat="output/Putah2JcaliP.cXX.txt"
  output:
    "output/Putah2JcaliP.assoc.txt"
  shell:
    "gemma -bfile calls/Jcali_primary/Putah2JcaliP -k {input.mat} -lmm 2 -miss 0.2 -o Putah2JcaliP"


rule gemma_lmm_alt:
  input:
    bed="calls/Jcali_alt/Putah2JcaliAlt.bed",
    bim="calls/Jcali_alt/Putah2JcaliAlt.bim",
    fam="calls/Jcali_alt/Putah2JcaliAlt.fam", #AFTER EDIT
    mat="output/Putah2JcaliAlt.cXX.txt"
  output:
    "output/Putah2JcaliAlt.assoc.txt"
  shell:
    "gemma -bfile calls/Jcali_alt/Putah2JcaliAlt -k {input.mat} -lmm 1 -miss 0.2 -maf 0.2 -o Putah2JcaliAlt"





# ========= LD =========

rule genoLD_all_2Jcali_primary:
  input:
    "calls/Jcali_primary/Putah_variant_filtered.vcf.gz"
  output:
    "Putah_Gloc_all_2Jcali_primary.geno.ld"
  conda:
    "/home/jgroh/heterodichogamy/conda_envs/vcftools.yaml"
  shell:
    """
    vcftools --gzvcf {input} \
        --chr JAKSXK010000007.1 --from-bp 31330000 --to-bp 31410000 \
        --maf 0.1 \
        --remove-indv JHIN_PC_040 --remove-indv JHIN_PC_060 \
        --geno-r2 --ld-window-bp-min 20 \
        --out Putah_Gloc_all_2Jcali_primary
    """

rule genoLD_gg_2Jcali_primary:
  input:
    "calls/Jcali_primary/Putah_variant_filtered.vcf.gz"
  output:
    "Putah_Gloc_protandrous_2Jcali_primary.geno.ld"
  conda:
    "/home/jgroh/heterodichogamy/conda_envs/vcftools.yaml"
  shell:
    """
    vcftools --gzvcf {input} --keep PutahCreek_gg_ids.txt \
        --remove-indv JHIN_PC_040 --remove-indv JHIN_PC_060 \
        --maf 0.05 \
        --chr JAKSXK010000007.1 --from-bp 31330000 --to-bp 31410000 \
        --geno-r2 --ld-window-bp-min 20 \
        --out Putah_Gloc_protandrous_2Jcali_primary
    """


rule genoLD_all_2Jcali_alt:
  input:
    "calls/Jcali_alt/Putah_variant_filtered.vcf.gz"
  output:
    "Putah_Gloc_all_2Jcali_alt.geno.ld"
  conda:
    "/home/jgroh/heterodichogamy/conda_envs/vcftools.yaml"
  shell:
    """
    vcftools --gzvcf {input} \
        --chr JAKSXL010000006.1 --from-bp 30467523 --to-bp 30570347 \
        --maf 0.1 \
        --remove-indv JHIN_PC_040 --remove-indv JHIN_PC_060 \
        --geno-r2 --ld-window-bp-min 20 \
        --out Putah_Gloc_all_2Jcali_alt
    """

rule genoLD_gg_2Jcali_alt:
  input:
    "calls/Jcali_alt/Putah_variant_filtered.vcf.gz"
  output:
    "Putah_Gloc_protandrous_2Jcali_alt.geno.ld"
  conda:
    "/home/jgroh/heterodichogamy/conda_envs/vcftools.yaml"
  shell:
    """
    vcftools --gzvcf {input} --keep PutahCreek_gg_ids.txt \
        --remove-indv JHIN_PC_040 --remove-indv JHIN_PC_060 \
        --maf 0.05 \
        --chr JAKSXL010000006.1 --from-bp 30467523 --to-bp 30570347 \
        --geno-r2 --ld-window-bp-min 20 \
        --out Putah_Gloc_protandrous_2Jcali_alt
    """
